autokeras/graph.py

```diff
@@ -279,20 +279,34 @@ def _get_loss(self):
 
     def _compile_keras_model(self, hp, model):
         # Specify hyperparameters from compile(...)
-        optimizer_name = hp.Choice(
-            "optimizer",
-            ["adam", "sgd", "adam_weight_decay"],
-            default="adam",
-        )
+        # check if optimizer is set before applying default optimizer settings.
+        if "optimizer" in hp.values.keys():
+            optimizer_name = hp.get("optimizer")
+        else:
+            optimizer_name = hp.Choice(
+                "optimizer",
+                ["adam", "sgd", "adam_weight_decay"],
+                default="adam",
+            )
+            
         # TODO: add adadelta optimizer when it can optimize embedding layer on GPU.
-        learning_rate = hp.Choice(
-            "learning_rate", [1e-1, 1e-2, 1e-3, 1e-4, 2e-5, 1e-5], default=1e-3
-        )
+        # check if learning rate is set before applying default learning rate settings.
+        if "learning_rate" in hp.values.keys():
+            learning_rate = hp.get("learning_rate")
+        else:
+            learning_rate = hp.Choice(
+                "learning_rate", [1e-1, 1e-2, 1e-3, 1e-4, 2e-5, 1e-5], 
+                default=1e-3,
+            )
 
         if optimizer_name == "adam":
             optimizer = keras.optimizers.Adam(learning_rate=learning_rate)
         elif optimizer_name == "sgd":
             optimizer = keras.optimizers.SGD(learning_rate=learning_rate)
+        elif optimizer_name == "rmsprop":
+            optimizer = keras.optimizers.RMSprop(learning_rate=learning_rate)
+        elif optimizer_name == "adadelta":
+            optimizer = keras.optimizers.Adadelta(learning_rate=learning_rate)
```

Comment:
Is there any specific reason why only these two solvers are added?

https://keras.io/api/optimizers/

If adding `Adadelta`, why not adding `Adagrad`, too?

Reply:
In fact, this is not my major purpose to add optimizers of adam series. It may just because I had seen adadelta in the comments of codes.

For RMSprop, that is mainly for personal preference. Besides variations of adam optimizers, optimizers implemented in Keras are SGD, RMSprop, and Ftrl according to official document. I'm wondering if the last one may be preferred for some other programmer.

For more general request of custom optimizer, it still need to refer to my previous comment: 
Serialization of optimizer settings.
But such serialization will treat learning rate as one parameter within optimizer settings, 
and you will need to implement your own optimizer as subclass of Keras optimizer.
Further more, learning rate scheduler might also be considered.

A possible reference or example for serialization is loss or metric, which implement serialization in engine/head.py.
Optimizer may need to follow such step in long term, but it may not be suitable to implement in the same place as loss and metric since it is reasonable to assume one model/graph can only have one universal optimizer, but one model/graph may have multiple heads. Maybe I am wrong here.

and optimizer serialization may not only put extra codes in utils/type.py like this simple way:
```
# leave loss for example
from tensorflow.keras.losses import Loss

LossType = Union[str, Callable, Loss]

# just share the same type definition with loss
from tensorflow.keras.optimizers import Optimizer

OptimizerType = Union[str, Callable, Optimizer]
```

The true purpose of this request may need more effort to accomplish though.

Reply:
Hey!
Would you be willing to work on adding functionality for passing loss and optimizer in the form of tunable hyper parameters as arguments for automodel?

Reply:
well, it is possible if overwriting optimizer can be avoided, but something weird happened for real code testing above.